### ISSUE TITLE
Race Condition in WSGIResource

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#9300]
|[<img alt="arnimarj's avatar" src="https://avatars.githubusercontent.com/u/143150?s=50" width="50" height="50">](https://github.com/arnimarj)| @arnimarj reported|
|-|-|
|Trac ID|trac#9300|
|Type|defect|
|Created|2017-10-02 20:00:23Z|

I believe the fix for [#8692](https://github.com/twisted/twisted/issues/8692) has revealed a race in the class `_WSGIResponse`.

`notifyFinished()` is called on the underlying request object when request has finished, but that fact is never used by pending writes scheduled by `_WSGIResponse`. It uses `callFromThread()` and `blockingCallFromThread()` from write `write()` and `run()` respectively and if they are schedule after `notifyFinished()` triggers they try to write to an already-closed transport.

Both `wsgiError()` and `wsgiWrite()` need to respect the `_requestFinished` flag.

The use of the `_requestFinished` in the `run()` function is also suspect since it is modified in the reactor thread and read from a background thread (the one calling `run()`).

I've attached an example which reveals the first race.

Attachments:

* [twisted_bug.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d8c/d8cda6680f79990514b867f4e66e5df7e9429120/43e701b1c0089ae485c20aa3d32aa83f8fb5918d.py) (876 bytes) - added by arnimarj on 2017-10-02 20:01:44Z - A small example revealing the race. Requires httpbin and requests.

<details><summary>Searchable metadata</summary>

```
trac-id__9300 9300
type__defect defect
reporter__arnimarj arnimarj
priority__normal normal
milestone__None None
branch__ 
branch_author__ 
status__new new
resolution__None None
component__web web
keywords__None None
time__1506974423482665 1506974423482665
changetime__1511724910565923 1511724910565923
version__None None
owner__arnimarj arnimarj

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="arnimarj's avatar" src="https://avatars.githubusercontent.com/u/143150?s=50" width="50" height="50">](https://github.com/arnimarj)<a name="note_3"></a>|@arnimarj set status to `closed`|
|-|-|

https://github.com/twisted/twisted/pull/907

**Comment by twisted-trac** [comment]
|[<img alt="arnimarj's avatar" src="https://avatars.githubusercontent.com/u/143150?s=50" width="50" height="50">](https://github.com/arnimarj)|@arnimarj set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="arnimarj's avatar" src="https://avatars.githubusercontent.com/u/143150?s=50" width="50" height="50">](https://github.com/arnimarj)|@arnimarj set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban set owner to @arnimarj|
|-|-|

Thanks for working on this.

The unit test looks like an integration/functional tests to me, but without docstring is hard to tell what is the scope of that test.

See my comment at https://github.com/twisted/twisted/pull/907#pullrequestreview-79035340

Please consider adding some docstrings to the test and/or splitting it into unit tests.

Thanks again!

**Comment by twm** [comment]
I used to see this traceback all the time, but haven't for ages. It looks like this was probably adequately addressed by #9300, so I'm going to close this issue. Feel free to reopen if I'm wrong about that, and thank you for all the work on this!
